Documentation/design-docs/event-counter.md

```diff
@@ -39,28 +43,60 @@ We believe adding some new top-level types will satisfy these requests:
+
+I think there is still an argument to be made that standardization could benefit libraries that are merely exporting the data for storage in a time series database because rendering isn't their concern. However these tools are going to care about data storage size. If a few lines of code to discriminate two cases is going to distinguish counters that produce 5 numbers and counters that produce 1, I think any good export tool is going to want that savings.
 
 ### Canonicalizing a single value output per counter ###
 
-For EventCounter and PollingCounter we expect simple viewers to use the display name as-is and use the value for 'Mean'. For AggregatingEventCounter and AggregatingPollingCounter, we expect simple viewers to display the display name with " / Min" appended after it. The value should be computed by reading the Mean statistic and multiplying it by the number of measurement intervals per minute. For example if the counter had display name "Exceptions Thrown", value 2, and Interval=1sec the viewer should display "Exceptions Thrown / Min" with value 120.
+For EventCounter and PollingCounter we expect simple viewers to use the display name as-is and use the value for 'Mean'. For IncrementingEventCounter and IncrementingPollingCounter, we expect simple viewers to display the display name with " / Min" appended after it. The display value should be computed by reading the 'Value' statistic and multiplying it by the number of measurement intervals per minute. For example if the counter had display name "Exceptions Thrown", value 2, and IntervalSec=1.01 the viewer should display "Exceptions Thrown / Min" with value 118.8.
 
 
-### EventStream format
+### Metadata
 
-We should add "DisplayName" and "Metadata" to the payload fields. We still need to define exactly what the encoding of Metadata is, but at minimum assume that it contains the counter type as a well-known constant.
+I added fields for DisplayName and CounterType directly to the payload. Vance suggested adding a 'Metadata' field, which I would be OK with, but I see no benefit to using it for fields we can anticipate will exist in advance, and we already have strongly typed APIs that generate their values. In the future if we added an API that let EventCounter consumers set arbitrary key value pairs on the counter, that seems like the data we'd want to encode in a Metadata string.
```

Comment:
The goal of the MetaData field is to allow new Meta-data to be defined by a user of the EventCounter API in cooperation with the viewer.   For example,   Lets imagine that the viewer wishes have some counters with  'per second' and others are 'per  min'.   If EventCounter ALREADY had a payload field  'MetaData'  field that was a even list of strings (representing key-value pairs).  

which is defined as space delimited key=value pairs, a user of EventCounter could have an PREDEFINED method which the user could do 

myCounter.AddMetaData("units, sec");

Which ends up emitting 'MetaData : [ "units" , "sec" } 

And then the viewer to look at this meta-data and if it recognizes 'units' and 'sec' then it would know what to do (and display as a /Sec rather than '/Min' default.  

Notice no changes were needed to EventCounter to do this.   This is important because otherwise we would need and update to EventCounter, which means an update to the runtime, which is often a big deal.  

Having strongly typed properties for well known ones is fine.   Things that are 'required' in the payload make sense to  have their own field in the payload.  Things that are optional (they effect the default that the viewer chooses, and the human probably has a way of overriding), probably belong in meta-data (like default units).

Reply:
I think we agree on the general idea (we might disagree on which specific metadata elements belong on the strongly-typed vs. weakly-typed side of the divide, but we can tackle that as the need arises). I didn't add an AddMetaData() API based on our email conversation where you suggested it was premature to add it.

Reply:
To have the value of allowing more data to be added without runtime updates,, something like AddMetaData needs to be added before we finally ship 3.0.

Reply:
I'd like for this informal spec to capture our consolidated thinking about what is in scope for 3.0 so we lock in our progress and there is no confusion about what we've agreed to vs. what is still under discussion. Assuming no objections about the AddMetaData API (I have none at least) then we should update the spec with a new PR.

That probably also means that I should raise now that I think the timescale units deserve a strongly typed property unlike the example you gave above. My proposal of the bar we use to decide strong/loose typing:
1) If we at Microsoft are unable to predict the need for it => AddMetaData()
2) If we believe a property usage is likely to be rare, or that most viewers wouldn't use it => AddMetaData()
3) Otherwise (predictable+common usage) => strongly typed property.
Happy to hear if anyone feels differently. My rationale is that things that aren't strongly typed are harder to discover in docs/intellisense, less likely be used uniformly, prone to typos and prone to more disagreement on what correct usage is. It also has a more uniform API usage where all properties can be set together:

```
    IncrementingEventCounter foo = new IncrementingEventCounter("bla", this)
    {
        DisplayName = "Bla Bla",
        DisplayRateTimeScale = 1 // number of seconds
    }
```

These all seem like desirable properties for the common+predictable case.

@sywhang - Could I hand off ownership of maintaining the current state of this spec to you? It just means that decisions from discussions get captured as PRs like what I have been doing up until now. You may also find you want to make proposals of your own or raise issues as you make more discoveries in your EventCounter work.

Reply:
@noahfalk Sure, I'd be happy to take over :) 

Just to continue the discussion here -- I agree with Noah that having a strongly typed for well-known metadata that we know precisely how they should be used is the better thing to do. 

For `AddMetaData`, should the API signature look like: `public void AddMetaData(object key, object value)` and just serialize these key/value in the payload by calling `ToString`?

Reply:
I am OK with Noah's suggestion above (we add 'AddMetaData' and we have strongly typed methods for any that we feel are common and we know about).

Note that there is a corresponding issue on how it is encoded in the packet (do we make more properties for the event or do we dump them into a meta-data string)   

I would argue that clients of this data, would prefer predictability (the data always comes in the same way).     Here is a strawman

1. Meta-data we consider 'essential' (most/all viewers will want to manipulate it), basically the ones we have already identified) can be properties of the event.  
2. Anything that is optional (that is anything we did not anticipate in V1), ends up in the meta-data field.

Reply:
> For `AddMetaData`, should the API signature look like: `public void AddMetaData(object key, object value)` and just serialize these key/value in the payload by calling `ToString`?

I suggest AddMetaData(string,string). The (object,object) version gives the illusion we can handle anything when in reality many types won't serialize to a string in a useful manner. I'd opt for a more limited abstraction that reliably handles what you give it vs. an abstraction that appears more powerful but doesn't work well on many inputs.

> Here is a strawman ...

Sounds good to me

Reply:
Agreed.   The goal here is NOT to get into the business of serializing things, but rather provide the simple possible mechanism for transferring a very limited amount of information from the counter to the viewer in a way that can be extended over time.    I considered just have a single string, but decided there was value in providing some structure the string is a list of (key value)s

Reply:
Thanks. I'll update the docs with what we just talked about above.